### PR TITLE
Avoid division by zero in hybrid point encoding

### PR DESCRIPTION
**Pull request by botovq**
In hybrid and compressed point encodings, the form octet contains a bit
of information allowing to calculate y from x.  For a point on a binary
curve, this bit is zero if x is zero, otherwise it must match the
rightmost bit of of the field element y / x.  The existing code only
considers the second possibility. It could thus incorrecly fail with a
division by zero error as found by Guido Vranken's cryptofuzz.

This commit adds a few explanatory comments to oct2point. The only
actual code change is in the last hunk which adds a BN_is_zero(x)
check to avoid the division by zero.

##### Checklist
- [x] tests are added or updated

### CODE REVIEW

**Comment by botovq** [comment]
As mentioned in #15021, I am going to sign the CLA.

**Review by romen — APPROVED** [review]
LGTM

Thanks for also including a test!

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
Waiting for the CLA to be signed.

File: test/ectest.c [hunk 1]
```diff
@@ -1083,6 +1083,46 @@ static int char2_field_tests(void)
     BN_free(yplusone);
     return r;
 }
+
+static int hybrid_point_encoding_test(void)
```
Comment (slontis) [inline test/ectest.c:1087]: Shouldnt there be a negative test also, for a bad encoding?

Comment (botovq) [reply test/ectest.c:1087]: I have tweaked the test so it exercises the newly added y_bit check.

**Review by romen — APPROVED** [review]
Even better, good point @slontis!

Formally reapproved.

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
Have you submitted the CLA already?

**Comment by t8m** [comment]
This will need make update after rebasing before merge as it touches FIPS module sources.

**Comment by botovq** [comment]
@t8m No, not yet. As mentioned in the linked issue, it might take a week
or two before I get access to a printer or scanner. I'll do it as soon
as I can.

Do I need to do 'make update' or is this something the person who merges 
will do?

**Comment by romen** [comment]
> This will need make update after rebasing before merge as it touches FIPS module sources.

@t8m Does this mean that we need to rescope what is considered beta1 blocker?
Up till now this sounded like a bug fix, so something that could be fixed even during the beta period (and back ported to 1.1.1).
Does the new FIPS source checking mechanism change this?

**Comment by t8m** [comment]
> Does the new FIPS source checking mechanism change this?

I do not think so. We never said that post beta1 no FIPS relevant source changes are allowed. I suppose we might want to restrict changes to them some time later during the beta cycle but it should be definitely allowed for some while after beta1.

**Comment by t8m** [comment]
> Do I need to do 'make update' or is this something the person who merges will do?

I think it can be done by the person who merges but of course feel free to do it yourself if you want. It should also fix the check_update test failure.

File: test/ectest.c (hunk 1 above)
Comment (slontis) [reply test/ectest.c:1087]: Thanks.

**Comment by kaduk** [comment]
> it might take a week or two before I get access to a printer or scanner.

I know that people with neither have been able to do some kind of electronic-only thing in the past, but I don't handle CLA processing so I don't know the details of how :(

**Comment by romen** [comment]
@botovq I will deal with the `make update` to update the FIPS checksum during merge, as soon as the other tests are finished.

**Comment by romen** [comment]
Merged to `master` with: 

- 56f0237938 Avoid division by zero in hybrid point encoding
- e70abb8b4c Test oct2point for hybrid point encoding of (0, y)
- f0f4a46c4f FIPS checksums update


Thanks @botovq for the contribution and @guidovranken for raising the issue!

**Comment by romen** [comment]
@botovq I tried force pushing to your branch to see if GitHub would recognize the PR as `merged` rather than `closed`.

Unfortunately I pushed to `master` first, so the test was not conclusive.
Sorry for the noise!

**Comment by richsalz** [comment]
> access to a printer or scanner

If you have a digitized copy of your signature already, free PDF tools can add it into an existing PDF. Or word-processing programs can also do it.  That might helps.

**Pull request closed without merging**